cpp/include/cuml/experimental/fil/detail/infer_kernel/cpu.hpp

```diff
@@ -108,33 +109,45 @@ void infer_kernel_cpu(forest_t const& forest,
 
     for (auto row_index = start_row; row_index < end_row; ++row_index) {
       for (auto tree_index = start_tree; tree_index < end_tree; ++tree_index) {
-        auto tree_output = std::conditional_t<has_vector_leaves,
+        auto tree_output  = std::conditional_t<has_vector_leaves,
                                               typename node_t::index_type,
                                               typename node_t::threshold_type>{};
-        if constexpr (has_nonlocal_categories) {
-          tree_output = evaluate_tree<has_vector_leaves>(
-            forest.get_tree_root(tree_index), input + row_index * col_count, categorical_data);
+        auto leaf_node_id = index_type{};
```

Comment:
Why do we need a separate variable here? Can't this be `tree_output` with appropriate modification of the type at line 112?

Reply:
In this context, `infer_type == infer_kind::leaf_id` is not constexpr. This condition needs to be evaluated at runtime. This is why we have two calls to `evaluate_tree_dispatch` below.

Reply:
Ah, you're quite right! And that answers my question about the indexing as well. Unfortunate, but I don't see any way around it. Thanks!